Support inline diagnostics in CommandReturnObject

adrian-prantl opened the pull request:
and implement them for dwim-print (a.k.a. `p`) as an example.

The next step will be to expose them as structured data in SBCommandReturnObject.

adrian-prantl commented:
Side note: I also think that it should be possible to change CommandObjectExpression to just put the error into CommendReturnObject instead of printing it to the error stream directly. But that's for a future patch.

jimingham commented on lldb/include/lldb/Expression/DiagnosticManager.h:
```diff
@@ -23,35 +23,6 @@
-    bool hidden = false;
-    bool in_user_input = false;
-  };
-  /// Contains {{}, 1, 3, 3, true} in the example above.
```
There are two bools in the structure but only one in the listed contents.  This should have two "true"-s, right?

jimingham replied:
Oh, wait, should be "I don't actually know what hidden means" and "false"...

jimingham commented on lldb/include/lldb/Utility/Status.h:
```diff
@@ -26,6 +27,35 @@ class raw_ostream;
+    bool hidden = false;
+    bool in_user_input = false;
+  };
+  /// Contains {{}, 1, 3, 3, true} in the example above.
```
This doesn't help me understand what "hidden" means.

adrian-prantl replied:
I updated the documentation, thanks for pointing this out!

jimingham commented on lldb/source/Commands/CommandObjectDWIMPrint.cpp:
```diff
@@ -188,12 +188,14 @@ void CommandObjectDWIMPrint::DoExecute(StringRef command,
     ValueObjectSP valobj_sp;
     std::string fixed_expression;
 
+    result.SetUserInput(expr);
```
This is at least an oddly named API, since it's not at all what the user input.  For instance, I might very well have run:

dwim-print -a 1 -l swift -- some expression

but this API is going to claim "some expression" is the user input.

jimingham commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -3180,15 +3180,24 @@ void CommandInterpreter::IOHandlerInputComplete(IOHandler &io_handler,
+    // Display any inline diagnostics first.
+    if (!result.GetImmediateErrorStream() &&
+        GetDebugger().GetShowInlineDiagnostics() &&
+        line.find('\n') == std::string::npos) {
```
When would you have a `\n` in a command string?

adrian-prantl replied:
In a multiline expression for example.

jimingham commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -3180,15 +3180,24 @@ void CommandInterpreter::IOHandlerInputComplete(IOHandler &io_handler,
+      PrintCommandOutput(io_handler, diags, true);
+    }
+
+    // Display any STDOUT/STDERR _prior_ to emitting the command result text.
```
Did this comment move too far?  I'm pretty sure GetProcessOutput() is the call that displays STDOUT/STDERR.

adrian-prantl replied:
Maybe? I don't know the answer to this, but this isn't something that I changed in this patch.

jimingham replied:
Yes, it definitely should bo before GetProcessOutput.  Your patch moved it from before that call to after both that call and your added text.

jimingham replied:
Still need to fix this comment.

jimingham commented on lldb/source/Interpreter/CommandReturnObject.cpp:
```diff
@@ -112,8 +113,46 @@ void CommandReturnObject::SetError(Status error) {
 void CommandReturnObject::SetError(llvm::Error error) {
-  if (error)
+  // Retrieve any diagnostics.
+  error = llvm::handleErrors(std::move(error), [&](ExpressionErrorBase &error) {
```
Why is this data structure called ExpressionErrorBase?  That's a clearly inappropriate name for errors in a generic CommandObject, and seeing it here is confusing.

adrian-prantl replied:
I renamed it to DiagnosticError and moved it to Utility.

jimingham commented on lldb/source/Interpreter/CommandReturnObject.cpp:
```diff
@@ -112,8 +113,46 @@ void CommandReturnObject::SetError(Status error) {
+}
+
+llvm::StringRef
+CommandReturnObject::GetInlineDiagnosticsData(unsigned indent,
```
This API seems wrong to me.  You have the CommandReturnObject being made up for some command that got executed, so I should know what that command was.  Then I have another API where I can pass in (if I feel like it) an entirely unrelated string as the "command".  Then I have some m_user_input which is some subset of what the user actually input for the command.
Maybe I'm missing something, but this is confusing to me.

adrian-prantl replied:
Agreed. I made it slightly less weird by moving the indent computation into CommandInterpreter, but I still can't say I like it.

jimingham commented:
I find this API confusing, and it seems very focused on pointing out errors in commands that have some part of their text that gets submitted to a compiler.

The other obvious candidate for this feature is a general command parser implementation that handles:

(lldb) process launch --no-such-option -arch x86_64
                                          ^
                                           | unrecognized option '--no-such-option'

But it's not entirely clear to me how this scheme would support that...

adrian-prantl commented:
@jimingham I changed the API to just record the offset of the expression in the command instead of the expression itself, which should make this all look a bit more logical.

jimingham commented on lldb/include/lldb/Utility/DiagnosticsRendering.h:
```diff
@@ -0,0 +1,79 @@
+  /// Contains "use of undeclared identifier 'foo'" in the example above.
+  std::string message;
+  /// Contains the fully rendered error message, without "error: ",
+  /// but including the source context.
```
What does this contain for "inline display"?

adrian-prantl replied:
@jimingham I don't understand the question. Concretely, in the example `rendered` is
```
 <user expression 0>:1:3: use of undeclared identifier 'foo'
 ///   1 + foo
 ///       ^~~
```
this is what a client can use if they choose to *not* show diagnostics inline.

jimingham replied:
I was confused because it didn't seem like there was a way to get the inline rendering if you wanted to use that, for instance if you were writing your own driver.   Presumably there's no way to do that?  That set me to wondering whether `is_user_input = true` would affect what you see in rendered?  
You are using "the fully rendered error message" to mean "the one not rendered for inline display".  That wasn't clear to me.

BTW, I don't actually think it's necessary to have a way to fetch the inline rendered form (properly offset for the prompt string, etc.)  The possibility set me wondering, however.

adrian-prantl replied:
There is no "fetching of the inline rendered form". Either you use the prerendered non-inline form or you have to render the message yourself using all the details in the struct. The point is to allow different drivers to render errors in a way that makes sense for them.

jimingham replied:
You have a bool saying whether this can be used in inline displays, which leads to the question "can I get you to make that for me here".  That the answer to that is no is fine, butsaying that explicitly would have reduced confusion.  No biggie...

jimingham commented on lldb/include/lldb/Utility/DiagnosticsRendering.h:
```diff
@@ -0,0 +1,79 @@
+  static char ID;
+};
+
+const char *ExpressionResultAsCString(lldb::ExpressionResults result);
```
There's a version of this that produces "human readable strings" out of an ExpressionResults (in ErrorMessages.{h, cpp}).  It would make more sense for this function to live alongside that one.

adrian-prantl replied:
I did not know about this file!

jimingham commented on lldb/include/lldb/Utility/DiagnosticsRendering.h:
```diff
@@ -0,0 +1,79 @@
+
+const char *ExpressionResultAsCString(lldb::ExpressionResults result);
+
+llvm::raw_ostream &PrintSeverity(Stream &stream, lldb::Severity severity);
```
This one should also live wherever the ExpressionResults renderer is.

jimingham commented:
You didn't introduce the situation where we have two separate "ExpressionResults -> String" API's in different places, but you did move one, so maybe it's time to put them together?  I don't have a strong opinion where they should go, except together if possible...

I'll still be curious to see how this works for any non-expression based command errors.  But ultimately if all it works for is expressions, it will still be doing a good thing, so...

adrian-prantl committed:
Support inline diagnostics in CommandReturnObject

and implement them for dwim-print (a.k.a. `p`) and the command option
parser (in one place thus far) as an example.

The next step will be to expose them as structured data in
SBCommandReturnObject.

JDevlieghere commented on lldb/source/Interpreter/Options.cpp:
```diff
@@ -19,12 +19,45 @@
 
+namespace lldb_private {
+
+/// An llvm::Error that represents an optino parsing diagnostic.
```
```suggestion
/// An llvm::Error that represents an option parsing diagnostic.
```

jimingham commented on lldb/include/lldb/Utility/Args.h:
```diff
@@ -38,19 +38,23 @@ class Args {
     bool IsQuoted() const { return quote != '\0'; }
     char GetQuoteChar() const { return quote; }
+    std::optional<uint16_t> GetPos() const { return column; }
+    uint16_t GetLength() const { return ref().size(); }
```
These are transitory object, made by a command when executed and destroyed when it is done.  So I don't think we need to worry about how big they are.

adrian-prantl replied:
1. This is the column position of something a human typed. Even uint8_t is a stretch.
2. In Clang and LLVM debug info metadata we (somewhat consistently) use uint16_t for column counts.

JDevlieghere replied:
I was talking specifically about `GetLength`, not `GetPos`. It makes sense that the column info is a `uint16_t`, but the length of the argument shouldn't be unnecessarily truncated (imho).

jimingham commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -2078,6 +2078,11 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
 
     ElapsedTime elapsed(execute_time);
     cmd_obj->SetOriginalCommandString(real_original_command_string);
+    pos = real_original_command_string.rfind(remainder);
```
IIUC, the indent gets incremented twice here.  The first time is here where you're just conveniently setting it to "past the command name" since commands don't get that part.  Then later if there's an error you bump the indent to the actual error.  That seems sufficiently tricky to warrant a comment?

jimingham commented on lldb/test/Shell/Commands/command-options.test:
```diff
@@ -0,0 +1,11 @@
+# RUN: echo quit | %lldb -o "log enable -x" \
```
Can you write one of these tests where the wrong option isn't the first?  From the standpoint of "things you could get wrong", "only working for the first option" seems a likely failure mode.

adrian-prantl replied:
Thanks! Indeed, there's bug there.

jimingham commented:
Except for the test request, this LGTM.

adrian-prantl commented on lldb/include/lldb/Interpreter/CommandReturnObject.h:
```diff
@@ -160,6 +166,9 @@ class CommandReturnObject {
   StreamTee m_err_stream;
+  std::vector<DiagnosticDetail> m_diagnostics;
+  StreamString m_diag_stream;
+  std::optional<uint16_t> m_diagnostic_indent;
```
Because DiagnostcDetails are created in places (e.g., the expression evaluator) that don't have any information about driver details like the indent.
